change various easyblock deriving from CMakeMake to use 'build_type' easyconfig parameter

Flamefire opened the pull request:
This replaces raw usages of `-DCMAKE_BUILD_TYPE=...` in other ECs

I also added a commit which renames the option to build_type because the base class ConfigureMake has this and having both might be confusing to users. This lets the last one win, in this case CMake

boegel commented on easybuild/easyblocks/o/opencv.py:
```diff
@@ -90,9 +91,6 @@ def prepare_step(self, *args, **kwargs):
     def configure_step(self):
         """Custom configuration procedure for OpenCV."""
 
-        if 'CMAKE_BUILD_TYPE' not in self.cfg['configopts']:
```
Shouldn't we add this check whether `CMAKE_BUILD_TYPE` is already specified` in `configopts` in `CMakeMake`?

Flamefire replied:
Not required. This is added first so configopts can overwrite it

boegel replied:
OK

boegel commented on easybuild/easyblocks/p/psi.py:
```diff
@@ -56,6 +56,7 @@ def __init__(self, *args, **kwargs):
         self.psi_srcdir = None
         self.install_psi_objdir = None
         self.install_psi_srcdir = None
+        self.cfg['build_type'] = 'Release'
```
Same issue w.r.t. default in `eb -a`, change via `extra_options`?

boegel commented on easybuild/easyblocks/p/psi.py:
```diff
@@ -143,9 +144,7 @@ def configure_step(self):
-                self.cfg.update('configopts', ' -DCMAKE_BUILD_TYPE=EasyBuildRelease')
-            else:
-                self.cfg.update('configopts', ' -DCMAKE_BUILD_TYPE=Release')
+                self.cfg['build_type'] = 'EasyBuildRelease'
```
This should be in the `if`?

Flamefire replied:
It is, the else was removed

boegel commented on easybuild/easyblocks/d/dirac.py:
```diff
@@ -54,7 +54,7 @@ def configure_step(self):
 
         self.cfg['separate_build_dir'] = True
-        self.cfg.update('configopts', "-DENABLE_MPI=ON -DCMAKE_BUILD_TYPE=release")
+        self.cfg['build_type'] = 'release'  # TODO: Verify if this is really lowercase
```
As far as I know the values passed to CMake options are case-insensitive (but you should confirm that)

Flamefire replied:
Not always. If it is used directly casing does matter (in a strequal e.g.)
I verified the other uses but can't for this one as access to source requires a license

Flamefire commented on easybuild/easyblocks/d/dolfin.py:
```diff
@@ -54,6 +54,8 @@ def __init__(self, *args, **kwargs):
         self.saved_configopts = None
         self.cfg['separate_build_dir'] = True
+        # run cmake in debug mode
+        self.cfg['build_type'] = 'Debug'
```
Yes. I opened an issue requesting clarification why this was added

boegel replied:
see #1920

Flamefire replied:
From the discussion I conclude that this was indeed a mistake. I'm currently building it with Release instead and will keep that if that works

Flamefire replied:
DOLFIN-2018.1.0.post1-foss-2018a-Python-3.6.4.eb build successfully with build_type=Release

Flamefire commented on easybuild/easyblocks/n/netcdf.py:
```diff
@@ -68,10 +68,7 @@ def configure_step(self):
-                self.cfg.update('configopts', '-DCMAKE_BUILD_TYPE=DEBUG ')
-            else:
-                self.cfg.update('configopts', '-DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_C_FLAGS_RELEASE="-DNDEBUG " ')
+            self.cfg['build_type'] = 'Debug' if self.toolchain.options.get('debug', None) else 'Release'
```
Cmake does add it by default. Can double check

boegel replied:
You mean it adds it when `-DCMAKE_BUILD_TYPE=Release` is used, not always, right?

Flamefire replied:
Short answer: Correct

To be exact I meant that it adds it to CMAKE_C/CXX_FLAGS_RELEASE by default which is used for the release build type. Yes there is a difference but doesn't matter for "make" as the generator.

Flamefire replied:
Verified by checking output of a test build: `-- CMAKE_C_FLAGS_RELEASE:   -O3 -DNDEBUG`

Flamefire commented on easybuild/easyblocks/t/trilinos.py:
```diff
@@ -98,11 +98,8 @@ def configure_step(self):
-        else:
-            self.cfg.update('configopts', "-DCMAKE_BUILD_TYPE:STRING=RELEASE")
+        # release or debug gversion
+        self.cfg['build_type'] = 'Debug' if self.toolchain.options.get('debug', None) else 'Release'
```
@boegel I think it makes sense to do this by default. Is the toolchainopt debug used for anything else?

boegel replied:
We may use it in some places to inject `-g` into `$CXXFLAGS` & co for example.

Always switching to `-DDCMAKE_BUILD_TYPE=Debug` when the `debug` toolchain option is enabled makes sense I guess...

Flamefire replied:
The question would be: How to handle 'debug' vs 'build_type'? My proposal: if build_type is not None, it overwrites 'debug'. Else it is used as it is here: Debug or Release
But this borders into the discussion in https://github.com/easybuilders/easybuild-easyblocks/pull/911#issuecomment-564466693: What for projects failing to work in "Release" mode? TBH I think this is a non-issue as optflags are usually used so only difference is (almost always) the added `-DNDEBUG`

To be conservative one could treat debug=None as not doing anything which makes it:
- if build_type is not None, use it
- else if debug is not None, use Release or Debug
- don't use anything

But then: Which ECs use `debug: False` explicitly? So I'd treat debug=None as false and if anything breaks, set it in the EC (again because I see very small chances for anything in ECs break)

Flamefire replied:
On the other hand current behavior for e.g. gromacs and netcdf is that toolchainopts has preference over build_type. Maybe use that?

Flamefire commented on easybuild/easyblocks/t/trilinos.py:
```diff
@@ -98,11 +98,8 @@ def configure_step(self):
-        else:
-            self.cfg.update('configopts', "-DCMAKE_BUILD_TYPE:STRING=RELEASE")
+        # release or debug gversion
+        self.cfg['build_type'] = 'Debug' if self.toolchain.options.get('debug', None) else 'Release'
```
@boegel I think it makes sense to do this by default. Is the toolchainopt debug used for anything else?

Flamefire replied:
@boegel I think it makes sense to do this by default. Is the toolchainopt debug used for anything else?

Flamefire commented on easybuild/easyblocks/c/clang.py:
```diff
@@ -96,6 +96,7 @@ def __init__(self, *args, **kwargs):
         self.llvm_obj_dir_stage2 = None
         self.llvm_obj_dir_stage3 = None
         self.make_parallel_opts = ""
+        self.cfg['build_type'] = 'Release'
```
Ok. Does it still need to by python 2 compatible?

boegel replied:
Yes, EasyBuild is still compatible with Python 2.6, 2.7 and 3.5+ (see https://easybuild.readthedocs.io/en/latest/Python-2-3-compatibility.html)

Flamefire commented:
Rebased and updated please reverify. I'd also like to get the behavior with toolchainopts.debug defined here. See https://github.com/easybuilders/easybuild-easyblocks/pull/1921#discussion_r366776798 My current stance would be that if toolchainopts.debug==True then build_type is set to Debug. That would match similar uses in existing ECs except that it still can be overwritten by passing it in configopts

Flamefire commented:
I ingested this into #1929 as changes requested there create conflicts with these changes and the scope became hard to tell apart. please continue discussion there
